pihole

```diff
@@ -192,6 +192,7 @@ Time:
     fi
   else
     # Enable Pi-hole
+    killall -q pihole-reenable
```

Comment:
Wouldn't it make sense to add `-w` option as well, so `restartDNS` is assured to not run until `pihole-reenable` has been terminated?
```
killall -qw pihole-reenable
```

Reply:
Yes, though I am not sure this is the way to go when I read what the [killall(1) man-page](https://www.man7.org/linux/man-pages/man1/killall.1.html) has to say about the `-w` option:
> killall may wait forever if the signal was ignored, had no effect, or if the process stays in zombie state.

That sounds like another problem waiting to happen, no?

Reply:
Reading up on how bash responds to the `SIGTERM` signal it appears that it [does not forward the signal to child processes](https://unix.stackexchange.com/a/196053/392028), if I am understanding it correctly. In other words, if the `pihole-reenable` process has started a `pihole enable` child process when it receives a `SIGTERM` signal from `killall`, the `pihole enable` child process is allowed to complete what it is doing. It is not being terminated abruptly.

If that is true, then at least this PR does not make the situation worse than it already is :-)

Reply:
This is probably a candidate for a new github issue. I believe it can be simulated like this:
```
pihole disable && (pihole enable & pihole enable &)
```

And this fails sometimes. Here is one case of a failure:
```
pi@pihole:~ $ pihole disable && (pihole enable & pihole enable &)
  [i] Disabling blocking
  [✓] Reloading DNS service
  [✓] Pi-hole Disabled
pi@pihole:~ $   [i] Enabling blocking
  [i] Enabling blocking
mv: cannot move '/etc/pihole/black.list.bck' to '/etc/pihole/black.list': No such file or directory
  [✓] Reloading DNS service
  [✓] Pi-hole Enabled
  [✓] Reloading DNS service
  [✓] Pi-hole Enabled

pi@pihole:~ $
```

So to summarize, yes there is a potentiel issue. However, this issue is already present before this PR.

Reply:
I just thought because killing existing re-enable jobs was the whole purpose of this PR but currently one cannot be sure how far this killing went, so cynical spoken: Nothing has changed 😉, however of course practically this will work in ~99% of cases where it is killed while executing `sleep`.

The question is if the concurrent re-enable job really hangs, is it better to let it hang and go on (I didn't track down what would happen) or hang as well and allow the user to recognise that something is not in order?

> Reading up on how bash responds to the SIGTERM signal it appears that it does not forward the signal to child processes, if I am understanding it correctly. In other words, if the pihole-reenable process has started a pihole enable child process when it receives a SIGTERM signal from killall, the pihole enable child process is allowed to complete what it is doing. It is not being terminated abruptly.

Bash definitely forwards SIGTERM, SIGINT and of course SIGKILL to child processes, as long as they are called in foreground:
```
2020-05-30 11:35:08 root@micha:/tmp# cat parent
#!/bin/bash
echo Started parent
/tmp/child
echo Stopped child, stopping parent, goodbye!
2020-05-30 11:38:13 root@micha:/tmp# cat child
#!/bin/bash
echo Started child
for i in {1..10}
do
        sleep 1
        echo Child running for $i seconds...
done
echo Stopping child ordinary after $i seconds
2020-05-30 11:38:16 root@micha:/tmp# ./parent &> log &
[1] 7543
2020-05-30 11:38:35 root@micha:/tmp# killall -w parent
[1]+  Terminated              ./parent &> log
2020-05-30 11:38:39 root@micha:/tmp# cat log
Started parent
Started child
Child running for 1 seconds...
Child running for 2 seconds...
Child running for 3 seconds...
Child running for 4 seconds...
Child running for 5 seconds...
Child running for 6 seconds...
2020-05-30 11:38:41 root@micha:/tmp# killall -w parent
parent: no process found
```
When calling the child like `/tmp/child &` it indeed continues in background, ignoring ALL signals to parent.
This btw shows that output of killall should be muted, STDOUT as well as STDERR, for "Terminated" and especially "not found" prints?

To make the kill more forcing, SIGKILL could be sent, which cannot be ignored by processes:
```
killall -qws SIGKILL pihole-reenable
```
No problem and faster for `sleep` but hard-killing `pihole enable` might have negative side effects as well.
_______________
So after all I would indeed go with `-w` and default SIGTERM signal. Disabling Pi-hole for a while with this command is mostly done manually from console, so users can react to a hanging command. While "sleep" can be assumed failsafe and fast when being killed, `pihole enable` indeed takes a noticeable time, so it makes sense to wait for it stopping. But even without it, this PR solves the issues in 99% of cases, so nothing urgent 😉.

Best would be actually, if `pihole-reenable` is running, if one could kill the `sleep` command only and leave `pihole enable` to be done by the existing `pihole-reenable` instance. However, complicated to to all the related checks and kill the precise `sleep` only, via PID or such 😄.

Reply:
> Best would be actually, if pihole-reenable is running, if one could kill the sleep command only and leave pihole enable to be done by the existing pihole-reenable instance.

Agree. And I believe this is, indeed, what this PR achieves.

I am not sure the parent/child example is correct though as the child process loses stdout when the parent process is terminated.

Without redirecting stdout this is the output from one console:
```
$ ./parent &
[1] 15479
$ Started parent
Started child
Child running for 1 seconds...
Child running for 2 seconds...
Child running for 3 seconds...
Child running for 4 seconds...
Child running for 5 seconds...
Child running for 6 seconds...
Child running for 7 seconds...
Child running for 8 seconds...
Child running for 9 seconds...
Child running for 10 seconds...
Stopping child ordinary after 10 seconds
```

And the output from another console after starting the parent process:
```
$ killall -v parent
Killed parent(15479) with signal 15
$
```

Reply:
tracing the experiment gives a wealth of information:
```
strace -f ./parent
```

Reply:
To clarify; when the `pihole-reenable` script receives `SIGTERM` ([in non-interactive mode](https://lists.gnu.org/archive/html/bug-bash/2020-03/msg00006.html)) it will terminate. If if happens when `sleep x` is running it means that the `pihole-reenable` script does not call `pihole reenable` (good).

If the child process `sleep x` is running, it will sleep until it's done with that. Should be ok, as long as you don't reach pid_max...
If the child process `pihole enable` is running, it will keep doing whatever it needs to do until it is done. (good)

Reply:
> I am not sure the conclusion from the parent/child example is correct, though, because the child process loses stdout when the parent process is terminated.

Ah you're right. just checked and indeed the child continues when running parent in foreground and killen from second shell 👍. Good to know indeed!

- So this means, when concurrent `pihole-reenable` is still executing `sleep`, it is terminated completely and the PR will do exactly what it aims.
- But if it has already called `pihole enable`, this will continue without us recognising it, even with `killall -w`.
- However the by far longest part of `pihole enable` is reloading the pihole-FTL service, which is done as last step. So when concurrent `pihole enable` is called, nearly immediately it sets `BLOCKING_ENABLED=true`, which makes the second `pihole enable` skipping pihole-FTL reload: https://github.com/pi-hole/pi-hole/blob/903808a47726cad878da329f17d529341df1c514/pihole#L201-L204
This is good since this is the only part where concurrent file writes and such could mess things up.
- So indeed the chance that any concurrency issues occur is very minimal with this PR, and `-w` practically has no effect 👍.

Reply:
The `pihole-reenable` script is called with `nohup` and `STDIN`, `STDOUT` and `STDERR`redirected to /dev/null: https://github.com/pi-hole/pi-hole/pull/2887/files#diff-808056da3f34bcc1499322fe165a334dR167

The `&>` redirect form appears to be  a bash4 thing according to https://en.wikipedia.org/wiki/Redirection_(computing)#Redirecting_to_and_from_the_standard_file_handles

With this PR applied I do not see any "Terminated" message when performing these commands:
```
pihole disable 30s
pihole enable
pihole disable
sleep 30
pihole status
```

Reply:
`cmd &>/dev/null` == `cmd >/dev/null 2>&1` on bash, yes.

No sure if setting STDIN to /dev/null has any effect here, however I just wanted to call sleep exactly like `pihole-reenable.sh` is. And as can be seen redirecting all file descriptors + set to background + nohup is not enough to mute the terminate print, as it is not directly an output of the command, but seems to be sent to the originating terminal directly somehow. Not sure  why `disown` or wrapping the killall into a command group with SDTERR + SDTOUT redirection helps, missing some insights into bash job handling.

However, you're right no output with the real scripts :+1:, probably because the running `pihole-reenable.sh` instance does not belong to the same shell, respectively has no parent shell anymore at all (previous `pihole` instance exited already).

Reply:
I guess the difference in behaviour regarding the "Terminated" messages could be an interactive/non-interactive thing. Good point about STDIN, I just left it as it was to make the change as simple as possible.

Anyway, thank you very much for the review and testing!